pkgs/desktops/lomiri/qml/lomiri-ui-toolkit/default.nix

```diff
@@ -0,0 +1,337 @@
+
+  src = fetchFromGitLab {
+    owner = "ubports";
+    repo = "development/core/lomiri-ui-toolkit";
+    rev = finalAttrs.version;
+    hash = "sha256-RH9xhbII2GfpxPg6QcPrBFIoI1/Ol7WmNdd0j9FkNs8=";
+  };
+
+  outputs = [ "out" "dev" ];
+
+  patches = [
+    # Upstreaming effort for these two patches: https://gitlab.com/ubports/development/core/lomiri-ui-toolkit/-/merge_requests/44
+    (fetchpatch {
+      name = "0001-lomiri-ui-toolkit-fix-tests-on-qt-5.15.4.patch";
+      url = "https://salsa.debian.org/ubports-team/lomiri-ui-toolkit/-/raw/1ad650c326ba9706d549d1dbe8335c70f6b382c8/debian/patches/0001-fix-tests-on-qt-5.15.4.patch";
+      hash = "sha256-Y5HVvulR2760DBzlmYkImbJ/qIeqMISqPpUppbv8xJA=";
+    })
+    (fetchpatch {
+      name = "0002-lomiri-ui-toolkit-fix-tests-on-qt-5.15.5.patch";
+      url = "https://salsa.debian.org/ubports-team/lomiri-ui-toolkit/-/raw/03bcafadd3e4fda34bcb5af23454f4b202cf5517/debian/patches/0002-fix-tests-on-qt-5.15.5.patch";
+      hash = "sha256-x8Zk7+VBSlM16a3V1yxJqIB63796H0lsS+F4dvR/z80=";
+    })
+  ];
+
+  postPatch = ''
+    patchShebangs documentation/docs.sh tests/
+
+    substituteInPlace tests/tests.pro \
+      --replace "\''$\''$PYTHONDIR" "$dev/${python3.sitePackages}"
+
+    for subproject in po app-launch-profiler lomiri-ui-toolkit-launcher; do
+      substituteInPlace $subproject/$subproject.pro \
+        --replace "\''$\''$[QT_INSTALL_PREFIX]" "$out" \
+        --replace "\''$\''$[QT_INSTALL_LIBS]" "$out/lib"
+    done
+
+    # Install apicheck tool into bin
+    substituteInPlace apicheck/apicheck.pro \
+      --replace "\''$\''$[QT_INSTALL_LIBS]/lomiri-ui-toolkit" "$out/bin"
+
+    # Causes redefinition error with our own fortify hardening
+    sed -i '/DEFINES += _FORTIFY_SOURCE/d' features/lomiri_common.prf
+
+    # Reverse dependencies (and their reverse dependencies too) access the function patched here to register their gettext catalogues,
+    # so hardcoding any prefix here will make only catalogues in that prefix work. APP_DIR envvar will override this, but with domains from multiple derivations being
+    # used in a single application (lomiri-system-settings), that's of not much use either.
+    # https://gitlab.com/ubports/development/core/lomiri-ui-toolkit/-/blob/dcb3a523c56a400e5c3c163c2836cafca168767e/src/LomiriToolkit/i18n.cpp#L101-129
+    #
+    # This could be solved with a reference to the prefix of whoever requests the domain, but the call happens via some automatic Qt / QML callback magic,
+    # I'm not sure what the best way of injecting that there would be.
+    # https://gitlab.com/ubports/development/core/lomiri-ui-toolkit/-/blob/dcb3a523c56a400e5c3c163c2836cafca168767e/src/LomiriToolkit/i18n_p.h#L34
+    #
+    # Using /run/current-system/sw/share/locale instead of /usr/share/locale isn't a great
+    # solution, but at least it should get us working localisations
+    substituteInPlace src/LomiriToolkit/i18n.cpp \
+      --replace "/usr" "/run/current-system/sw"
```

Comment:
do you know how other desktop environments on nixos handle this?

Reply:
Closest functional & situational equivalent in a similar DE seems to be `ki18n` for KDE, it uses [a less-hardcoded way of finding a matching locale prefix](https://invent.kde.org/frameworks/ki18n/-/blob/3f1ffd29e9df5c1f62b12dd6f8c1f9f0b4ced5a9/src/i18n/kcatalog.cpp#L217-233)? Assuming it works for KDE (haven't tried), I can try to port its method to LUITK and see if that improves things.

Otherwise, our [Cinnamon DE](https://github.com/NixOS/nixpkgs/blob/f320b0d46ee33f521f50d7c6885605714210e5c9/pkgs/desktops/cinnamon/cinnamon-common/default.nix#L163) [packaging](https://github.com/NixOS/nixpkgs/blob/f320b0d46ee33f521f50d7c6885605714210e5c9/pkgs/desktops/cinnamon/cinnamon-screensaver/default.nix#L87) also patches such hardcoded locale dirs to `/run/current-system/sw/share/locale`.

Reply:
[0001-LomiriToolkit-i18n-Locate-locale-via-QStandardPaths-.patch.txt](https://github.com/NixOS/nixpkgs/files/13004227/0001-LomiriToolkit-i18n-Locate-locale-via-QStandardPaths-.patch.txt) is a quick attempt at porting the ki18n way of doing this to LUITK. I think it should work, but it'll need [a setup hook like in ECM](https://github.com/NixOS/nixpkgs/blob/ca012a02bf8327be9e488546faecae5e05d7d749/pkgs/development/libraries/kde-frameworks/extra-cmake-modules/setup-hook.sh#L60-L98) to bring the paths of QML dependencies into `XDG_DATA_DIRS` (worst case, `/run/current-system/sw/share` is already in there so it would be on-par with current submission), and I'm not sure this is acceptable for upstream as-is.

There might also be [an existing-but-seemingly-underused way of making sure the domains are properly bound](https://gitlab.com/ubports/development/core/lomiri-system-settings/-/blob/d3f6014276c63639c850c6fae4175e34de21d33a/src/qml/MainWindow.qml#L76-77) which might be lower-maintenance, but I haven't played around with that yet.

This is definitely doable, but I'll lack the free time to properly finish this up with proper testing of everything, setup hooks & upstreaming now that new university semester has started. I hope the current submission is flawed but good enough for now.